src/Kaixa.groovy

```diff
@@ -2321,20 +2329,62 @@ public class Kaixa {
       cachedPasswords.put(name, password);
     }
 
-    // Wait for discovery page to load
-    Kaixa.waitForElementPresent('#idp_1001962798_button');
-    Kaixa.click('#idp_1001962798_button');
-
-    // Wait for the page to load
-    Kaixa.waitForElementVisible('#username');
-    Kaixa.waitForElementVisible('#password');
-
-    // Add credentials
-    Kaixa.typeInto('#username', username);
-    Kaixa.typeInto('#password', password);
-
-    // Click "submit"
-    Kaixa.click('.btn-primary');
+		boolean isHarvardKey  = obj.has('isHarvardKey') && obj.getBoolean('isHarvardKey');
+		boolean needs2FA = isHarvardKey && obj.has('needs2FA') && obj.getBoolean('needs2FA');
+
+		if (expectDiscoveryPage) {
+			// Wait for discovery page to load
+			Kaixa.waitForAtLeastOneElementPresent([
+				'#idp_1001962798_button', // HarvardGuest
+				'#idp_1824601020_button', // HarvardKey
+			]);
```

Comment:
This would be the feature you could use to check if there's a discovery page. This function returns which one appeared. You could wait for _either_ the discovery page or the login page and handle it appropriately